keps/sig-cli/kubectl-apply-prune-beta.md

```diff
@@ -0,0 +1,250 @@
+
+### Apply-Config Object
+To solve the first challenge, a concept of apply-config object is proposed. An apply-config object is an object that contains the GKN information for all the objects that are to be applied. The apply-config object is not generated by kubectl when applying is running. Instead, it should be generated by other tools such as kustomize, Helm. When generating resources to be consumed by `kubectl apply`, those tools are responsible to generate the associated apply-config object to utilize better pruning.
+
+An object is identified as an apply-config object when it has a special label `kubectl.kubernetes.io/apply-cfg: "true"`. The GKN information for all other objects are stored in annotations. Here is an apply-config object example.
+```
+apiVersion: v1
+kind: ConfigMap
+metadata:
+ name: name-foo
+ labels:
+   kubectl.kubernetes.io/apply-cfg: "true"
+ annotations:
+   kubectl.kubernetes.io/last-applied-objects: |
+     namespace1:group1:kind1:somename
+     namespace1:group2:kind2:secondname
+     namespace3:group3:kind3:thirdname
+```
+Note that the version is not necessary to find the object. The version of the resource can be selected by the client and available versions can be discovered dynamically
+
+The apply-config object is not limited to ConfigMap type. Any type of Object can be an apply-config object as long as it has the `apply-cfg` label. That means a custom resource object, such as an [Application](https://github.com/kubernetes-sigs/application) custom resource object, can also be used as an apply-config object when needed.
+
+The apply-config object could have an optional hash label `kubectl.kubernetes.io/apply-hash`, which is based on the content of annotation `kubectl.kubernetes.io/last-applied-objects`. This hash label is used for validation in kubectl. More details are in [How It Works](#how-it-works).
+
+### Child Objects
+The objects passed to apply will have annotations pointing back to the apply-config object for the same apply.
+```
+ annotations:
+    kubectl.kubernetes.io/applyOwner: uuid_of_the_apply_cfg_object 
+```
+The presence of this annotation is to confirm that a live object is safe to delete when it is to be pruned. A child object may be updated or touched by other users or applyOwner, in which case the annotation will be missing or different. In this scenario, it is safe to not delete the child object. This annotation is used to prevent deleting child objects in this scenario.
+
+### How it works
+Since the apply-config object is not generated by `kubectl`, other tools or users need to generate an apply-config object configuration before piping them into `kubectl apply`. For example, in following workflows, `kustomize` or `helm` need to generate the apply-config object. Moreover, every time those tools run, it should generate the apply-config object with the same name.
+```
+kustomize build . | kubectl apply --prune -f -
+# or
+helm template mychart | kubectl apply --prune -f -
```

Comment:
After running `helm template mychart | kubectl apply --prune -f -`, no apply-config will be generated by Helm. `helm template` is a "dry run" and the Release object is applied *after* the resources have been applied to the cluster. In other words, the release is recorded during a `helm install`, *after* the resources have been applied to the cluster. I don't understand how Helm is supposed to tie in here given the previous assumptions made above.

Reply:
The apply-config object should be passed to `kubectl apply` at the same time when applying resources. For example, `kustomize build .` has the output as
```
apply-config object
deployment object
...
```
I think `helm template` can also generate the apply-cfg object when rendering templates in a chart. Then users who pipe the output to `kubectl` can benefit the pruning functionality.

Reply:
It rather sounds like you're suggesting that you will change the way `kubectl` works to be solely aligned with `kustomize`, and other tools are just going to be required to adapt to `kustomize`'s pattern. Is that what you are saying?